src/eos-updater-poll-lan.c

```diff
@@ -0,0 +1,920 @@
+                             gboolean *valid,
+                             GError **error)
+{
+  if (!check_ostree_path (lan_data,
+                          g_hash_table_lookup (records, eos_avahi_v1_ostree_path ()),
+                          valid,
+                          error))
+    return FALSE;
+  if (!*valid)
+    // TODO: message
+    return TRUE;
+  return TRUE;
+}
+
+static ServiceValidResult
+txt_v1_handler_time_check (LanData *lan_data,
+                           EosServiceWithBranchFile *swbf,
+                           GHashTable *records)
+{
+  const gchar *dl_time;
+  g_autoptr(GDateTime) txt_utc = NULL;
+
+  dl_time = g_hash_table_lookup (records, eos_avahi_v1_branch_file_dl_time ());
+  switch (check_dl_time (lan_data, dl_time, &txt_utc))
+    {
+    case DL_TIME_OLD:
+      eos_service_with_branch_file_get_from_lan_data (swbf, lan_data);
+      return SERVICE_VALID;
+
+    case DL_TIME_NEW:
+      swbf->declared_download_time = g_steal_pointer (&txt_utc);
+      return SERVICE_VALID_MORE_CHECKS;
+
+    case DL_TIME_INVALID:
+      // TODO: message
+      return SERVICE_INVALID;
+
+    case DL_TIME_OUT_OF_RANGE:
+      // TODO: message
+      return SERVICE_INVALID;
+    }
+  // TODO: message
+  g_assert_not_reached ();
+  return SERVICE_INVALID;
+}
+
+static gboolean
+txt_v1_handler_checks (LanData *lan_data,
+                       EosServiceWithBranchFile *swbf,
+                       GHashTable *records)
+{
+  ServiceValidResult result;
+
+  result = txt_v1_handler_time_check (lan_data, swbf, records);
+  if (result == SERVICE_VALID_MORE_CHECKS)
+    {
+      const gchar *sha512sum;
+      const gchar *sha512sum_local;
+
+      sha512sum = g_hash_table_lookup (records, eos_avahi_v1_branch_file_sha512sum ());
+      sha512sum_local = lan_data->fetch_data->data->branch_file->contents_sha512sum;
+      if (g_strcmp0 (sha512sum_local, sha512sum) == 0)
+        eos_service_with_branch_file_get_from_lan_data (swbf, lan_data);
+      else
+        swbf->declared_sha512sum = g_strdup (sha512sum);
+      result = SERVICE_VALID;
+    }
+  if (result == SERVICE_VALID)
+    return TRUE;
+  return FALSE;
+}
+
+static gboolean
+txt_v1_handler (LanData *lan_data,
+                EosServiceWithBranchFile *swbf,
+                gboolean *valid,
+                GError **error)
+{
+  g_autoptr(GHashTable) records = NULL;
+  TxtRecordError txt_error = get_unique_txt_records (swbf->service->txt,
+                                                     &records,
+                                                     eos_avahi_v1_ostree_path (),
+                                                     eos_avahi_v1_branch_file_dl_time (),
+                                                     eos_avahi_v1_branch_file_sha512sum (),
+                                                     NULL);
+  if (txt_error != TXT_RECORD_OK)
+    {
+      // TODO: message
+      *valid = FALSE;
+      return TRUE;
```

Comment:
The commit simplifying it is https://github.com/kinvolk/eos-updater/commit/048017ac9f5b077b5600e05a73fbb90a353c072f